k2/csrc/fsa_algo.cu

```diff
@@ -540,4 +540,93 @@ void AddEpsilonSelfLoops(FsaOrVec &src, FsaOrVec *dest,
   }
 }
 
+// TODO(fangjun): compute arc_map
+Fsa Union(FsaVec &fsas, Array1<int32_t> *arc_map /*= nullptr*/) {
+  K2_CHECK_EQ(fsas.NumAxes(), 3);
+
+  ContextPtr &context = fsas.Context();
+  const int32_t *fsas_row_splits1_data = fsas.RowSplits(1).Data();
+  const int32_t *fsas_row_splits2_data = fsas.RowSplits(2).Data();
+  const int32_t *fsas_row_ids1_data = fsas.RowIds(1).Data();
+  const int32_t *fsas_row_ids2_data = fsas.RowIds(2).Data();
+  const Arc *arcs_data = fsas.values.Data();
+
+  int32_t num_fsas = fsas.Dim0();
+  int32_t num_states = fsas.TotSize(1);
+  int32_t num_arcs = fsas.TotSize(2);
+
+  // A new start state and a new final state are added (+2).
+  // The final state of each fsa is removed (-num_fsas)
+  int32_t num_out_states = num_states + 2 - num_fsas;
+  int32_t out_final_state = num_out_states - 1;
+
+  // For every fsa, a new arc is added from the new start state
+  // to its original start state (+num_fsas)
+  int32_t num_out_arcs = num_arcs + num_fsas;
```

Comment:
and here, if one of the input fsa is empty. That's why I strongly ask you add random tests and I think I have asked this too many times..